src/Streamly/Internal/Data/Fold/Type.hs

```diff
@@ -826,9 +1011,59 @@ shortest _f1 _f2 = undefined
 -- /Unimplemented/
 --
 {-# INLINE longest #-}
-longest :: -- Monad m =>
+longest :: Monad m =>
     Fold m x a -> Fold m x a -> Fold m x a
-longest _f1 _f2 = undefined
+longest (Fold stepL beginL doneL) (Fold stepR beginR doneR) =
+    Fold step begin done
+
+    where
+
+    begin = do
+        resL <- beginL
+        resR <- beginR
+        return
+            $ case resL of
+                    Partial sl ->
+                        Partial
+                            $ case resR of
+                                Partial sr -> LongestRunBoth sl sr
+                                Done br -> LongestRunLeft sl br
+                    Done bl -> first (LongestRunRight bl) resR
+
+    step (LongestRunBoth sL sR) a = do
+        resL <- stepL sL a
+        resR <- stepR sR a
+        case resL of
+            Partial sL1 ->
+                return
+                    $ Partial
+                    $ case resR of
+                        Partial sR1 -> LongestRunBoth sL1 sR1
+                        Done bR -> LongestRunLeft sL1 bR
+            Done bL ->
+                return
+                    $ case resR of
+                        Partial sR1 -> Partial $ LongestRunRight bL sR1
+                        Done bR -> Done bR
+    step (LongestRunLeft sL bR) a = do
+        resL <- stepL sL a
+        return
+            $ case resL of
+                Partial sL1 -> Partial $ LongestRunLeft sL1 bR
+                Done bL -> Done bL
+    step (LongestRunRight bL sR) a = do
+        resR <- stepR sR a
+        return
+            $ case resR of
+                Partial sR1 -> Partial $ LongestRunRight bL sR1
+                Done bR -> Done bR
+
+    --done (RunBoth sL sR) = do -- which one to pick ?
+    done (LongestRunBoth _ _) = undefined
```

Comment:
We should extract the result here. Since none has finished we can take the first one.